books/RayTracingTheNextWeek.html

```diff
@@ -2057,105 +2057,147 @@
 
-            data = stbi_load(
-                filename, &width, &height, &components_per_pixel, components_per_pixel);
+    #include <cstdlib>
+    #include <iostream>
 
-            if (!data) {
-                std::cerr << "ERROR: Could not load texture image file '" << filename << "'.\n";
-                width = height = 0;
+    class rtw_image {
+      public:
+        rtw_image() : data(nullptr) {}
+
+        rtw_image(const char* image_filename) {
+            // Loads image data from the given file name. If the RTW_IMAGES environment variable is
+            // defined, looks only in that directory for the image file. If RTW_IMAGES is not
+            // defined, this constructor will search for the specified image file first from the
+            // current directory, then in the images/ subdirectory, then the _parent's_ images/
+            // subdirectory, and then _that_ parent, on so on, for six levels up. If the image was
+            // not loaded successfully, width() and height() will return 0.
+
+            const auto imagedir = getenv("RTW_IMAGES");
+            if (imagedir) {
+                // If RTW_IMAGES is defined, then only use that value to locate the image file.
+                if (load(std::string(imagedir) + "/" + image_filename)) return;
+            } else {
+                // Hunt for the image file in some likely locations.
+                auto filename = std::string(image_filename);
+                if (load(filename)) return;
+                if (load("images/" + filename)) return;
+                if (load("../images/" + filename)) return;
+                if (load("../../images/" + filename)) return;
+                if (load("../../../images/" + filename)) return;
+                if (load("../../../../images/" + filename)) return;
+                if (load("../../../../../images/" + filename)) return;
+                if (load("../../../../../../images/" + filename)) return;
             }
 
-            bytes_per_scanline = bytes_per_pixel * width;
-        }
-
-        ~image_texture() {
-            delete data;
+            std::cerr << "ERROR: Could not load image file '" << image_filename << "'.\n";
+            std::cerr << "STBI: " << stbi_failure_reason() << '\n';
         }
 
-        virtual color value(double u, double v, const vec3& p) const override {
-            // If we have no texture data, then return solid cyan as a debugging aid.
-            if (data == nullptr)
-                return color(0,1,1);
+        ~rtw_image() { STBI_FREE(data); }
 
-            // Clamp input texture coordinates to [0,1] x [1,0]
-            u = interval(0,1).clamp(u);
-            v = 1.0 - interval(0,1).clamp(v);  // Flip V to image coordinates
+        bool load(const std::string filename) {
+            // Loads image data from the given file name. Returns true if the load succeeded.
+            auto components_per_pixel = bytes_per_pixel; // Components per pixel
```

Comment:
remove the comment, or make it clarifying somehow

Reply:
Consider creating a class member variable
`const int components_per_pixel = 3` below

Reply:
No, this is correct; the STBI interface is clumsy here. If the last argument is 0, then the library will determine the number of bytes per pixel and store it in the out (penultimate) parameter. If the last argument is 1, 2, 3 or 4, the load function will transform the pixel to that format (I, IA, RGB, or RGBA) on load. However, it still takes the penultimate parameter and sets it to the actual value from the file.

We basically always want RGB, so we don't care about this value. I'll just rename it `dummy` to try to alleviate some of the confusion.